src/services/goToDefinition.ts

```diff
@@ -159,16 +157,30 @@ export function getDefinitionAtPosition(program: Program, sourceFile: SourceFile
             break;
     }
 
-    if (node.kind === SyntaxKind.AwaitKeyword) {
-        const functionDeclaration = findAncestor(node, n => isFunctionLikeDeclaration(n));
-        const isAsyncFunction = functionDeclaration && some(functionDeclaration.modifiers, node => node.kind === SyntaxKind.AsyncKeyword);
-        return isAsyncFunction ? [createDefinitionFromSignatureDeclaration(typeChecker, functionDeclaration)] : undefined;
+    // for keywords related to function or method definitions
+    let findFunctionDecl: ((n: Node) => boolean | "quit") | undefined;
+    let checkFunctionDeclaration: ((decl: FunctionLikeDeclaration) => boolean) | undefined;
+    switch (node.kind) {
+        case SyntaxKind.ReturnKeyword:
+            findFunctionDecl = n => {
+                return isClassStaticBlockDeclaration(n)
+                    ? "quit"
+                    : isFunctionLikeDeclaration(n);
+            };
+            break;
+        case SyntaxKind.AwaitKeyword:
+            checkFunctionDeclaration = (functionDeclaration: FunctionLikeDeclaration) => some(functionDeclaration.modifiers, node => node.kind === SyntaxKind.AsyncKeyword);
+            findFunctionDecl = isFunctionLikeDeclaration;
+            break;
+        case SyntaxKind.YieldKeyword:
+            checkFunctionDeclaration = functionDeclaration => !!functionDeclaration.asteriskToken;
+            findFunctionDecl = isFunctionLikeDeclaration;
+            break;
     }
-
-    if (node.kind === SyntaxKind.YieldKeyword) {
-        const functionDeclaration = findAncestor(node, n => isFunctionLikeDeclaration(n));
-        const isGeneratorFunction = functionDeclaration && functionDeclaration.asteriskToken;
-        return isGeneratorFunction ? [createDefinitionFromSignatureDeclaration(typeChecker, functionDeclaration)] : undefined;
+    if (findFunctionDecl) {
+        const functionDeclaration = findAncestor(node, findFunctionDecl) as FunctionLikeDeclaration | undefined;
+        const isCorrectDeclaration = functionDeclaration && (!checkFunctionDeclaration || checkFunctionDeclaration(functionDeclaration));
+        return isCorrectDeclaration ? [createDefinitionFromSignatureDeclaration(typeChecker, functionDeclaration)] : undefined;

```

Comment:
I kind of find the indirection here hard to follow. I can see wanting to find some common logic, but I'd suggest just writing it out explicitly in each branch for now.

Reply:
Also, the idea that you *wouldn't* want to still jump to the nearest function even if it's not `async` or a generator is something I'm not so sure of. So the idea that you need a `checkFunctionDeclaration` to validate that you got a function of the right kind feels like it would break down unnecessarily during refactoring.